### ISSUE TITLE
Using parameter "all" for service "ask_for_update" gives error in v0.5.4

### ISSUE THREAD

**Issue by bcutter** [gluap/pyduofern-hacs#35]
Running v0.5.4 using
```
service: duofern.ask_for_update
data:
  all: true
```

gives this in the UI: `Error calling service duofern.ask_for_update. Unknown error`
and this in the HA log:

```
2023-04-02 18:38:51.208 ERROR (MainThread) [homeassistant.helpers.script.websocket_api_script] websocket_api script: Error executing script. Unexpected error for call_service at pos 1: local variable 'sets' referenced before assignment
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 451, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 684, in _async_call_service_step
    await service_task
  File "/usr/src/homeassistant/homeassistant/core.py", line 1745, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1788, in _execute_service
    await self._hass.async_add_executor_job(
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/duofern/__init__.py", line 152, in ask_for_update
    getDuofernStick(hass).command(device_id, 'getStatus')
  File "/usr/local/lib/python3.10/site-packages/pyduofern/duofern_stick.py", line 576, in command
    self.duofern_parser.set(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/pyduofern/duofern.py", line 770, in set
    logger.debug(sets.keys())  # join(" ", sort keys sets)
UnboundLocalError: local variable 'sets' referenced before assignment
2023-04-02 18:38:51.215 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [547173666864] Error handling message: Unknown error (unknown_error) from xxx.xxx.xxx.xxx (Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/111.0)
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/decorators.py", line 27, in _handle_async_response
    await func(hass, connection, msg)
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 646, in handle_execute_script
    await script_obj.async_run(msg.get("variables"), context=context)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 1524, in async_run
    await asyncio.shield(run.async_run())
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 409, in async_run
    await self._async_step(log_exceptions=False)
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 453, in _async_step
    self._handle_exception(
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 476, in _handle_exception
    raise exception
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 451, in _async_step
    await getattr(self, handler)()
  File "/usr/src/homeassistant/homeassistant/helpers/script.py", line 684, in _async_call_service_step
    await service_task
  File "/usr/src/homeassistant/homeassistant/core.py", line 1745, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1788, in _execute_service
    await self._hass.async_add_executor_job(
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/duofern/__init__.py", line 152, in ask_for_update
    getDuofernStick(hass).command(device_id, 'getStatus')
  File "/usr/local/lib/python3.10/site-packages/pyduofern/duofern_stick.py", line 576, in command
    self.duofern_parser.set(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/pyduofern/duofern.py", line 770, in set
    logger.debug(sets.keys())  # join(" ", sort keys sets)
UnboundLocalError: local variable 'sets' referenced before assignment
```

**Interesting part:**

```
  File "/config/custom_components/duofern/__init__.py", line 152, in ask_for_update
    getDuofernStick(hass).command(device_id, 'getStatus')
  File "/usr/local/lib/python3.10/site-packages/pyduofern/duofern_stick.py", line 576, in command
    self.duofern_parser.set(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/pyduofern/duofern.py", line 770, in set
    logger.debug(sets.keys())  # join(" ", sort keys sets)
UnboundLocalError: local variable 'sets' referenced before assignment
```

**Comment by bcutter** [comment]
Additional notes after a bit more of blackbox testing (to find a work around for updating all devices):

❓ Gives no error but likely does nothing (?):
```
service: duofern.ask_for_update
data: {}
```

❌ Gives an error (wrong syntax)
```
service: duofern.ask_for_update
data:
  device_id: all
```

❓ Gives no error and makes use of HA magic "all", but likely does nothing (according to debug logs which contain nothing):
```
service: duofern.ask_for_update
data:
  device_id:
    - all
```
❓ Gives no error and uses a group of devices (cover group), but likely does nothing (according to debug logs which contain nothing):
```
service: duofern.ask_for_update
data:
  device_id:
    - cover.all_covers
```

Before I update my automations.yaml and scripts.yaml the 7th time during the last weeks I want to be sure to have a working service (syntax) 😃

**Comment by gluap** [comment]
**Regarding the original error**

>   File "/usr/local/lib/python3.10/site-packages/pyduofern/duofern.py", line 770, in set
>     logger.debug(sets.keys())  # join(" ", sort keys sets)
> UnboundLocalError: local variable 'sets' referenced before assignment

You mentioned that you had bogous devices (ones that you didn't pair yourself) detected by duofern - I believe having a potentially non-existing device type triggered this. Sets were always set for "existing" duofern device types - therefore this didn't show up before. but the bogous devices seem to not match that description. I fixed duofern to not raise that error any more (will be in 0.5.4). you will likely be able to trigger this by running the update exclusively by updating one of the "bogous" devices. 

**Regarding "how to use all"**
Your initial attempt is fine.
```
service: duofern.ask_for_update
data:
  all: true
```
Is the yaml the UI also produces when checking the box and it works on my side. The error it triggered on your side because of the ghost devices should be fixed with 0.5.4. (can't test for lack of ghost devices).

**Regarding the logging:**

Already in 0.5.3 there should always be info level logs when running the ``ask_for_update``. They look like this:
```
2023-04-02 21:19:27.503 INFO (SyncWorker_7) [custom_components.duofern] asking ffffff for update
```
If they are not there the command didn't go out. This may have been caused by the duofern thread crashing due to the issue you describe above.

Nevertheless I added a bunch more verbosity to the update function - it might help in case we get another issue with bogous devices.

**Comment by bcutter** [comment]
Interesting. I used the service pointing at only one specific device_id with debug logging enabled and there is - no - log entry.
Will update to 0.5.5.

**Comment by bcutter** [comment]
Now with 0.5.5 I get this:

When running

**1. As intended**

```
service: duofern.ask_for_update
data:
  all: true
```
--> Absolutely NOTHING in debug log.

**2. As not intended**
```
service: duofern.ask_for_update
data: {}
```
--> gives `2023-04-02 23:58:02.891 WARNING (SyncWorker_0) [custom_components.duofern] Found no valid device ids???`

❓

**Comment by bcutter** [comment]
My fault. I used the "enable debug logging" from the GUI:
![grafik](https://user-images.githubusercontent.com/13799156/230184282-14e9676b-6c8e-4bc8-8420-ffa381a44c9e.png)

...which created no logs for duofern at all.

But using

```
service: logger.set_level
data:
  custom_components.duofern: debug
```

and firing
```
service: duofern.ask_for_update
data:
  all: true
```

afterwards gave the expected log output. (I used the same service to set it back to "warning" for duofern) Unfortunately it reminded me of my plenty ghost devices, I think I need a separate issue for them :-(

**So I can confirm the new "all" syntax is working. I also did not see this issue anymore since on 0.5.5.** So I closed this one.
